content/posts/2019/06/nat_traversal_2.eng.md

```diff
@@ -0,0 +1,69 @@
+
+## Blockchain & NAT 
+
+Libplanet is a library that makes blockchain technology easy to use in game development. Many blockchain implementations use [P2P]-type networks for communications between distributed [nodes], and so does Libplanet. Unlike other implementations, however, the blockchain application implemented through Libplanet is a game. Typically, these games run on personal computers, on non-stop consoles, and on personal devices such as smartphones, most of which do not have a separate, authorized IP on a NAT-configured network. So, applications running on these devices must pass through NAT for P2P communication.
+ 
+For this reason, Libplanet supports relaying using TURN from 0.2.0. And here's how it works.
+
+## Port Allocation
+
+The first step in relaying over the TURN is a step known as port allocation. Nodes that want to broadcast from outside of NAT send an allocation request (with the appropriate credentials, if necessary) to the TURN server. The request dialogue will be something like this.
+
+> Node: I'd like to receive a relay through an authorized IP and port. Please allot an appropriate IP and port.
+
+If the request is correct, the TURN server will select and open the appropriate IP and port to receive the connection depending on the settings, and send the following response.
+
+
+> TURN Server: Port assigned. The address to be relayed in the future is '54.12.1.3:65002'. (Nonce: 'xyz')
+
+This connection that requests port allocation is called *control connection* and it is used for communications between TURN server and the node. One caveat here is the nonce that the TURN server sends with its response. Nonce is unique for each control connection and the acquired nonce must be included in all requests of the control connection. (Or, you may receive a stale nonce error and request it again.)
+
+## Authorization Request and Approval
+
+With the assigned IP and port ('54.12.1.3:65002'), it would be great if other nodes were able to connect directly, but we need to go through one more step—Authorization Request. This is to request the TURN server to enable the allocated port connect to other IP nodes. Any connections that have not been requested for authorization are blocked, which means that the node you want to relay must have prior IP information of the node you want to connect to. 
+
+But in real life, many users don't know or care about their own authorized IP, so it's very difficult to communicate this information directly. To resolve this, use cases such as [WebRTC] often figure out the IP information of the node to access at the [signalling] stage. Meanwhile, Libplanet uses the STUN protocol to check if the IP is behind NAT, and if so, it will send the IP to other nodes along with the IP/port that is relayed as authorized IP. The nodes that receive this information go through the process of requesting permission first via an authorized IP before they can access the information that has been relayed.
```

Comment:
```suggestion
But in real life, many users don't know or care about their own public IP, so it's very difficult to communicate this information directly. To resolve this, use cases such as [WebRTC] often figure out the IP information of the node to access at the [signalling] stage. Meanwhile, Libplanet uses the STUN protocol to check if the IP is behind NAT, and if so, it will send the IP to other nodes along with the IP/port that is relayed as public IP. The nodes that receive this information go through the process of requesting permission first via an public IP before they can access the information that has been relayed.
```